src/operators.jl

```diff
@@ -360,204 +360,168 @@ end
 one_indexed(A) = all(x -> isa(x, Base.OneTo), Compat.axes(A))
 function Compat.LinearAlgebra.diagm(x::AbstractVector{<:AbstractVariableRef})
     @assert one_indexed(x) # Base.diagm doesn't work for non-one-indexed arrays in general.
-    diagm(0=>copyto!(similar(x, GenericAffExpr{Float64,eltype(x)}), x))
-end
-
-###############
-# The _multiply!(buf,y,z) adds the results of y*z into the buffer buf. No bounds/size
-# checks are performed; it is expected that the caller has done this, has ensured
-# that the eltype of buf is appropriate, and has zeroed the elements of buf (if desired).
-
-function _multiply!(ret::Array{T}, lhs::Array, rhs::Array) where T<:JuMPTypes
-    m, n = size(lhs,1), size(lhs,2)
-    r, s = size(rhs,1), size(rhs,2)
-    for i ∈ 1:m, j ∈ 1:s
-        q = ret[i,j]
-        _sizehint_expr!(q, n)
-        for k ∈ 1:n
-            tmp = convert(T, lhs[i,k]*rhs[k,j])
-            add_to_expression!(q, tmp)
-        end
+    diagm(0=>copyto!(similar(x, GenericAffExpr{Float64, eltype(x)}), x))
+end
+
+###############################################################################
+# Matrix/Vector Arithmetic with JuMP eltypes
+###############################################################################
+
+###############################################################################
+# convenience/utility definitions
+
+const GenericAffOrQuadExpr = Union{GenericAffExpr, GenericQuadExpr}
+
+densify_with_jump_eltype(x::AbstractMatrix) = convert(Matrix, x)
+
+function densify_with_jump_eltype(x::SparseMatrixCSC{V}) where {V <: AbstractVariableRef}
+    return convert(Matrix{GenericAffExpr{Float64, V}}, x)
+end
+
+# See https://github.com/JuliaLang/julia/pull/18218.
+_A_mul_B_eltype(::Type{R}, ::Type{S}) where {R, S} = typeof(one(R) * one(S) + one(R) * one(S))
+
+_A_mul_B_ret_dims(A::AbstractMatrix, B::AbstractVector) = (size(A, 1),)
+_A_mul_B_ret_dims(A::AbstractMatrix, B::AbstractMatrix) = (size(A, 1), size(B, 2))
+
+# Don't do size checks here; defer that to `_A_mul_B(A, B)`.
+function _A_mul_B_ret(A, B, dims...)
+    T = _A_mul_B_eltype(eltype(A), eltype(B))
+    ret = Array{T}(undef, _A_mul_B_ret_dims(A, B))
```

Comment:
Why not do `zeros(T, ...)` instead of `Array...` followed by `fill_with_zeros!` ?

Reply:
I just carried this pattern over from the old code. It does leave a nice path to generalize to non-`Array` `ret`s later, but that's probably premature. 

I could replace it with `zeros` if we want to make that change.

Reply:
`zeros` is broken for JuMP types because `zero` returns a mutable value.